packages/web-components/src/theme/set-theme.ts

```diff
@@ -3,34 +3,52 @@ import * as tokens from './design-tokens.js';
 
 const tokenNames = Object.keys(tokens) as (keyof Theme)[];
 
+const SUPPORTS_REGISTER_PROPERTY = 'registerProperty' in CSS;
+const SUPPORTS_ADOPTED_STYLE_SHEETS = 'adoptedStyleSheets' in document;
+const themeStyleSheet = new CSSStyleSheet();
+const themeStyleTextMap = new Map<Theme, string>();
+
 /**
  * Sets the theme tokens on defaultNode.
  * @param theme - Flat object of theme token values.
  * @internal
  */
 export const setTheme = (theme: Theme) => {
-  for (const t of tokenNames) {
-    let registered = false;
-
-    if ('registerProperty' in CSS) {
-      try {
-        CSS.registerProperty({
-          name: `--${t}`,
-          inherits: true,
-          initialValue: theme[t] as string,
-        });
-        registered = true;
-      } catch {
-        // Do nothing.
+  // Fallback to setting token custom properties on `<body>` element’s `style`
+  // attribute, only checking the support of  `document.adoptedStyleSheets`
+  // here because it has broader support than `CSS.registerProperty()`, which
+  // is checked later.
+  if (!SUPPORTS_ADOPTED_STYLE_SHEETS) {
+    setThemeFor(document.body, theme);
+    return;
+  }
+
+  if (!themeStyleTextMap.has(theme)) {
+    const tokenDeclarations: string[] = [];
+
+    for (const t of tokenNames) {
+      if (SUPPORTS_REGISTER_PROPERTY) {
+        try {
+          CSS.registerProperty({
+            name: `--${t}`,
+            inherits: true,
+            initialValue: theme[t] as string,
+          });
+        } catch {}
       }
+      tokenDeclarations.push(`--${t}: ${theme[t] as string};`);
     }
 
-    if (!registered) {
-      // TODO: Find a better way to update the values. Current approach adds
-      // lots of code to the `style` attribute on `<body>`. Maybe look into
-      // `document.adoptedStyleSheets`.
-      setThemeFor(document.body, theme);
-    }
+    themeStyleTextMap.set(theme, `body{${tokenDeclarations.join('')}}`);
```

Comment:
should this be `:root {}` instead of body?

Reply:
Good point! I updated this. I'm using `html` to avoid unnecessarily high specificity that comes with `:root`.